File: backend/src/main/java/com/main/Difficult/Services/ProductoService.java
package com.main.Difficult.Services;

import com.main.Difficult.Domain.producto.Producto;
import com.main.Difficult.Exceptions.NotFoundException;
import com.main.Difficult.Repositories.mongo.ProductoRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

@Service
public class ProductoService {
    @Autowired private ProductoRepository repoProductos;

    @Transactional(readOnly = true)
    public Producto findById(String idProducto){
        return repoProductos
                .findById(idProducto)
                .orElseThrow(()-> new NotFoundException("Producto Inexistente"));
    }
    @Transactional(readOnly = true)
    public List<Producto> buscarPor(String nombre, String pais, Integer puntaje){
        return repoProductos.buscarPor(nombre,pais,puntaje);
    }
}

File: backend/src/main/kotlin/aplicacion/servicios/ProductoService.kt
package aplicacion.servicios

import aplicacion.dominio.producto.DatosDeBusqueda
import aplicacion.dominio.producto.Producto
import aplicacion.excepciones.NotFoundException
import aplicacion.repositorios.RepoProductos
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
open class ProductoService {
    @Autowired private lateinit var repoProductos: RepoProductos
    @Transactional(readOnly = true)
    open fun findById(id: Long): Producto {
        return repoProductos.findById(id).orElseThrow {
            throw NotFoundException("Producto inexistente")
        }
    }
    @Transactional(readOnly = true)
    open fun buscarPor(datosDeBusqueda: DatosDeBusqueda): List<Producto> {
        return repoProductos.buscarPor(datosDeBusqueda)
    }
}

